cpp/arcticdb/processing/aggregation.cpp

```diff
@@ -261,55 +247,42 @@ namespace
                     using GlobalRawType = typename GlobalTypeDescriptorTag::DataTypeTag::raw_type;
                     using MaybeValueType = MaybeValue<GlobalRawType, T>;
                     auto prev_size = aggregated_.size() / sizeof(MaybeValueType);
                     aggregated_.resize(sizeof(MaybeValueType) * unique_values);
-                    auto col_data = input_column->column_->data();
                     auto out_ptr = reinterpret_cast<MaybeValueType*>(aggregated_.data());
                     std::fill(out_ptr + prev_size, out_ptr + unique_values, MaybeValueType{});
-                    entity::details::visit_type(input_column->column_->type().data_type(), [&input_column, &groups, &out_ptr, &col_data] (auto type_desc_tag) {
-                        using ColumnTagType = std::decay_t<decltype(type_desc_tag)>;
-                        using ColumnType =  typename ColumnTagType::raw_type;
-                        if constexpr(!is_sequence_type(ColumnTagType::data_type)) {
-                            auto lambda = [&col_data, &out_ptr, &groups](auto iter) {
-                                while (auto block = col_data.next<TypeDescriptorTag<ColumnTagType, DimensionTag<entity::Dimension::Dim0>>>()) {
-                                    auto ptr = reinterpret_cast<const ColumnType *>(block.value().data());
-                                    for (auto i = 0u; i < block.value().row_count(); ++i, ++ptr, ++iter) {
-                                        auto& val = out_ptr[groups[deref(iter)]];
-                                        if constexpr(std::is_floating_point_v<ColumnType>) {
-                                            const auto& curr = GlobalRawType(*ptr);
-                                            if (!val.written_ || std::isnan(static_cast<ColumnType>(val.value_))) {
-                                                val.value_ = curr;
-                                                val.written_ = true;
-                                            } else if (!std::isnan(static_cast<ColumnType>(curr))) {
-                                                if constexpr(T == Extremum::MAX) {
-                                                    val.value_ = std::max(val.value_, curr);
-                                                } else {
-                                                    val.value_ = std::min(val.value_, curr);
-                                                }
-                                            }
+                    details::visit_type(input_column->column_->type().data_type(), [&input_column, &groups, &out_ptr] (auto col_tag) {
+                        using col_type_info = ScalarTypeInfo<decltype(col_tag)>;
+                        if constexpr(!is_sequence_type(col_type_info::data_type)) {
+                            Column::for_each_enumerated<typename col_type_info::TDT>(*input_column->column_, [&groups, &out_ptr](auto enumerating_it) {
```

Comment:
This is so much clearer than what was there before, as well as helping people to avoid performance pitfalls. Very exciting!